### PR TITLE
LNS: Replace Web3 package with a custom minimal web3 client implementation

### PR DESCRIPTION
**Pull request by mainnet-pat**
Windows deterministic build does not work on my arm64 mac, so I could not test.
However, I checked that the problematic packages
- bitarray
- cytoolz
- lru-dict

are not required anymore

### CODE REVIEW

**Comment by EchterAgo** [comment]
@mainnet-pat I just invited you to my repo that has an automatic Windows & Linux build set up. You can just push what you want to build to the master branch and it will be uploaded to https://ec.loping.net/ when the build is finished. I've also pushed this change there already, so binaries or a build log with errors should be there soon.

**Comment by EchterAgo** [comment]
It still seems to want to install all 3 packages, `bitarray`, `cytoolz` and `lru-dict`

**Comment by EchterAgo** [comment]
Ah, you haven't updated the deterministic requirements.

**Comment by cculianu** [comment]
Thanks @EchterAgo  for getting involved.

> It still seems to want to install all 3 packages, bitarray, cytoolz and lru-dict

Darn. Sadly, all 3 of those packages are distributed on pypi as source-only, so they are a no-go for the current way that we build windows (we lack a MSVC compiler installed capable of building source packages). See this comment: https://github.com/Electron-Cash/Electron-Cash/pull/2414#issuecomment-1091055295

I suppose we could figure out how to build python and/or python packages ourselves within wine/mingw32 in some way (either using gcc or figuring out how to install MSVC inside wine?).

Another alternative is I could build binary wheels for those packages myself on my system here (or Axel can do it since he's smarter than me with this stiff) -- and we can just have the Windows wine build process download those `.whl` files that we pre-built and host, and solve the problem that way.

Thoughts?

**Comment by cculianu** [comment]
> Ah, you haven't updated the deterministic requirements.

Oh. So it may not need those packages. Gotcha.  I will update the requirements in a branch and try it out.

File: contrib/requirements/requirements.txt [hunk 1]
```diff
@@ -16,4 +16,6 @@ pathvalidate
 dnspython>=2.0
 
 # for LNS support
-web3
+eth_abi
+eth_utils
```
Comment (cculianu) [inline contrib/requirements/requirements.txt:20]: These pachage names are: `eth-abi` and `eth-utils`...

**Comment by cculianu** [comment]
Ok, I looked at the deps.  `bitarray` and `lru-dict` are no longer required. However,sadly, `eth-abi` depends on `eth-utils` which ultimately depends on `cytoolz` which itself is source-only distributed (thus not compilable/installable using our current Windows builder).

I am afraid this PR is insufficeint to get the (current) Windows builds working.

So we are forced to go to plan B or plan C or plan D:

Either:

- Get the windows builder actually able to compile native code for Python packages
- Just pre-compile the souce-dist-only packages on another system and host the compiled `.whl` files somewhere and have the windows build process download them explicitly and install them.
- Figure out a way to reduce dependencies further.

*sigh* I wish these EVM people would compile their own packages... and upload them to PyPI.

**Comment by mainnet-pat** [comment]
Damn, I thought I did a proper cleanup after removing web3. Let me think if I can avoid using eth-utils

**Comment by mainnet-pat** [comment]
> @mainnet-pat I just invited you to my repo that has an automatic Windows & Linux build set up. You can just push what you want to build to the master branch and it will be uploaded to https://ec.loping.net/ when the build is finished. I've also pushed this change there already, so binaries or a build log with errors should be there soon.

hey Axel, I have just pushed to the repo's master but do not see anything happening at https://ec.loping.net

**Comment by mainnet-pat** [comment]
This builds for me locally on mac os. Trying to fix build for linux in Axel's repo

**Comment by cculianu** [comment]
@mainnet-pat - I haven't reviewed/merged this yet because @EchterAgo  actually is working on getting the windows builder capable of building python packages itself using MSVC.. which is quite a feat because we need to get MSVC command-line build tools installed inside a docker VM using WINE.. it's nuts.

Axel says he thinks he almost has it working.. in which case we may just go with the initial approach of using straight-up `web3` (current master's approach).

We have hit this problem before in our windows builds and it severely limited what python packages we can use for Electron Cash. This time around if Axel solves the windows builder problem we may just go with the (simpler/shorter) version of your LNS code that just uses `web3` (current master)

anyway that's the status of what's going on.. so you know.

I do appreciate your taking the time to reduce the deps though.. and in case Axel doesn't manage to get the windows builder working.. we can fall back to this PR.

Thank you.

**Comment by EchterAgo** [comment]
I actually managed to build this with MSVC in Docker just now:

image

image

is there anything you want me to test on Windows?

I'm currently using the code here: https://github.com/EchterAgo/Electron-Cash/tree/msvc_test

I'll clean this up ASAP and push it to my build server. I hope it has enough disk space :\

**Comment by EchterAgo** [comment]
https://ec.loping.net/4.2.7-36-gc3c7d0806/ test build

**Comment by cculianu** [comment]
@EchterAgo You are a genius!! W00.  So we can just use `web3` directly and not have to re-import its code.  This liberated Electron Cash tremendously for not only smartBCH related plugins, but in general.  W00 h00!

As for testing it: I think it's sufficient just to do a lookup like you did, since all paths lead to basically the same code for extraction of smartBCH contracts.... so if that worked it all work! 🎉

**Comment by cculianu** [comment]
Closed this in favor of the approach used in #2420

**Pull request closed without merging**